src/gui/DatabaseWidget.cpp

```diff
@@ -650,6 +656,73 @@ void DatabaseWidget::openUrl()
     }
 }
 
+void DatabaseWidget::downloadFavicon()
+{
+#ifdef WITH_XC_NETWORKING
+    const QModelIndexList selected = m_entryView->selectionModel()->selectedRows();
+    if (selected.isEmpty()) {
+        return;
+    }
+
+    QList<Entry*> selectedEntries;
+    for (const QModelIndex& index : selected) {
+        selectedEntries.append(m_entryView->entryFromIndex(index));
+    }
+
+    if (selectedEntries.isEmpty()) {
+        return;
+    }
+
+    auto* iconDownloader = new IconDownloaderDialog(this);
+    connect(iconDownloader, SIGNAL(entryUpdated()), SIGNAL(databaseModified()));
+
+    if (selectedEntries.count() == 1) {
+        connect(iconDownloader,
+            SIGNAL(messageEditEntry(QString,MessageWidget::MessageType)),
+            SLOT(showMessage(QString,MessageWidget::MessageType)));
+        iconDownloader->downloadFavicon(m_db, selectedEntries.first());
+        return;
+    }
+
+    m_iconDownloaderState = IconDownloaderState::Downloading;
+    iconDownloader->raiseWindow();
+    iconDownloader->show();
+    iconDownloader->activateWindow();
+    iconDownloader->raise();
+    auto result = AsyncTask::runAndWaitForFuture(
+        [&]() { return iconDownloader->downloadAllFavicons(m_db, selectedEntries); });
+    Q_UNUSED(result);
+    m_iconDownloaderState = IconDownloaderState::Idle;
+#endif
+}
+
+void DatabaseWidget::downloadAllFavicons()
+{
+#ifdef WITH_XC_NETWORKING
+    if (m_iconDownloaderState == IconDownloaderState::Downloading) {
+        return;
+    }
+    m_iconDownloaderState = IconDownloaderState::Downloading;
+
+    auto* iconDownloader = new IconDownloaderDialog(this);
+    connect(iconDownloader, SIGNAL(entryUpdated()), SIGNAL(databaseModified()));
+
+    Group* currentGroup = m_groupView->currentGroup();
+    Q_ASSERT(currentGroup);
+    if (currentGroup) {
+        auto entries = currentGroup->entries();
+        iconDownloader->raiseWindow();
+        iconDownloader->show();
+        iconDownloader->activateWindow();
+        iconDownloader->raise();
+        auto result = AsyncTask::runAndWaitForFuture(
+            [&]() { return iconDownloader->downloadAllFavicons(m_db, entries); });
+        Q_UNUSED(result);
```

Comment:
All of this code should be stuffed into the "downloadAllFacivons" function itself.